src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +617,159 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
+
+	return OidIsValid(foreignKeyOid);
+}
+
+
+/*
+ * FindForeignKeyOidWithName searches the foreign key constraint with
+ * inputConstraintName in the given list of foreign key constraint OIDs.
+ * Returns the OID of the matching constraint. If there no matching constraint
+ * in the given list, then returns InvalidOid.
+ */
+static Oid
+FindForeignKeyOidWithName(List *foreignKeyOids, const char *inputConstraintName)
+{
+	Oid foreignKeyOid = InvalidOid;
+	foreach_oid(foreignKeyOid, foreignKeyOids)
 	{
-		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		char *constraintName = get_constraint_name(foreignKeyOid);
 
-		if (constraintForm->contype == CONSTRAINT_FOREIGN)
+		if (strncmp(constraintName, inputConstraintName, NAMEDATALEN) == 0)
 		{
-			systable_endscan(scanDescriptor);
-			heap_close(pgConstraint, NoLock);
-
-			return true;
+			return foreignKeyOid;
 		}
-
-		heapTuple = systable_getnext(scanDescriptor);
 	}
 
-	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, NoLock);
-
-	return false;
+	return InvalidOid;
 }
 
 
 /*
- * ConstraintIsAForeignKey returns true if the given constraint name
- * is a foreign key to defined on the relation.
+ * GetTableForeignConstraintOidsInternal takes in a relationOid, and returns
+ * the list of OIDs for foreign constraints that the relation with relationOid
+ * is involved according to "flags" argument. In that sense,
+ * "flags" should include either:
+ *  - "EXTRACT_REFERENCING_CONSTRAINTS" to get the ones in which our relation is
+ *    the "referencing" one, or
+ *  - "EXTRACT_REFERENCED_CONSTRAINTS" to get the ones in which our relation is
+ *    the "referenced" one.
+ * Self-referencing foreign keys will not be considered if "flags" includes
+ * EXCLUDE_SELF_REFERENCES as well.
  */
-bool
-ConstraintIsAForeignKey(char *constraintNameInput, Oid relationId)
+static List *
+GetTableForeignConstraintOidsInternal(Oid relationOid, int flags)
 {
+	AttrNumber attributeNumber = InvalidAttrNumber;
+
+	/* either of them should be passed */
+	Assert(flags & (EXTRACT_REFERENCING_CONSTRAINTS | EXTRACT_REFERENCED_CONSTRAINTS));
+
+	if (flags & EXTRACT_REFERENCING_CONSTRAINTS)
+	{
+		attributeNumber = Anum_pg_constraint_conrelid;
+
+		/* only one of them should be passed at a time */
+		Assert(!(flags & EXTRACT_REFERENCED_CONSTRAINTS));
+	}
+	else if (flags & EXTRACT_REFERENCED_CONSTRAINTS)
+	{
+		attributeNumber = Anum_pg_constraint_confrelid;
+	}
+
+	bool excludeSelfReference = (flags & EXCLUDE_SELF_REFERENCES);
+
+	List *foreignKeyOids = NIL;
+
 	ScanKeyData scanKey[1];
 	int scanKeyCount = 1;
 
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
+	/*
+	 * Note that since there is no index for the column Anum_pg_constraint_confrelid,
+	 * this function performs sequential scan on pg_constraint in that case.
+	 */
+	bool hasIndex = (attributeNumber == Anum_pg_constraint_conrelid) ? true : false;
```

Comment:
`hasIndex` -> `useIndex`

Also, maybe move `ConstraintRelidTypidNameIndexId` to a variable and  set it when we need to use